### ISSUE TITLE
Basic machine learning algorithms.

### ISSUE THREAD

**Issue by diffrxction** [dsc-tiu/open-source-ml-ai#2]
Here, you are required to give self implemented machine learning algorithms. This means, submitting code that works similar to the inbuilt machine learning algorithms or error metrics like RMSE, MAE, simple regressors and classifiers.

Folder structure to follow:
```
├── Self-implemented ML Algorithms
│   ├── Error metrics
│   │   ├──Mean Absolute Error  (Create sub-folder here)
│   │   ├──Root Mean Absolute Error (Create sub-folder here) and so on.
│   ├── ML Algorithms
│   │   ├──Linear Regression  (Create sub-folder here)
│   │   ├──Logistic Regression  (Create sub-folder here) and so on.
│   │   ├──Decision Trees(Create sub-folder here)
│   │   ├──Support Vector Machines(Create sub-folder here)
```

**P.S. If you use any datasets to check the correctness of your algorithms, then do include the links in the relevant markdown files.**

If you are interested, add a comment to this issue and I will allocate the issue to you.

Please see, you will be allotted a small portion of the topics mentioned above just so all interested people have a chance to contribute.
You can also try implementing some metric or algorithm that is not mentioned in this issue. Do so by mentioning what you plan on implementing in the comments below or creating a new issue and I will assign the issue to you.
Also notice the labels for each issue, that will give you an idea about what kind of work you are expected to do.

Cheers! 😄

**Comment by ThePsyk3y** [comment]
I would like to try to self implement the Logistic Regression ML Algorithm. Please allot that section to me.

**Comment by diffrxction** [comment]
> I would like to try to self implement the Logistic Regression ML Algorithm. Please allot that section to me.

@ThePsyk3y Sure, feel to free to proceed with the pull request and remember to follow the folder structure as shown above. All the best! :star:

**Comment by ArmaanSeth** [comment]
Hello @diffrxction  I would like to write code for Linear Regression, including both its implementation and by using sklearn model.

**Comment by diffrxction** [comment]
> Hello @diffrxction I would like to write code for Linear Regression, including both its implementation and by using sklearn model.

Hey @ArmaanSeth I was thinking if your could write that code from scratch along with some documentation instead of using the scikit learn in-built library, that would be a better contribution as it would give people an idea about the internal working of the algorithm. Let me know if that is possible for you and I will allot this issue to you alongside the importing of libraries with some code examples.

**Comment by ArmaanSeth** [comment]
@diffrxction Yes I would be happy in doing so.

**Comment by diffrxction** [comment]
> @diffrxction Yes I would be happy in doing so. Just want to ask in which format do you require the documentation, -will it a separate text document or -will it be just comments in the code.

I was going for one subfolder for your task with a relevant title. Then within that subfolder, add a .py or .ipynb file as is convenient for you along with a Markdown file (.md file) explaining and documenting the idea/working behind the code.

**Comment by ArmaanSeth** [comment]
> > @diffrxction Yes I would be happy in doing so. Just want to ask in which format do you require the documentation, -will it a separate text document or -will it be just comments in the code.
> 
> I was going for one subfolder for your task with a relevant title. Then within that subfolder, add a .py or .ipynb file as is convenient for you along with a Markdown file (.md file) explaining and documenting the idea/working behind the code.

Okay, I am on it.

**Comment by maheakdave** [comment]
Hello @diffrxction  bhaiya i would like to write code for decision trees from scratch and its documentation

**Comment by diffrxction** [comment]
> Hello @diffrxction bhaiya i would like to write code for decision trees from scratch and its documentation (both regressor and classifier)

Sure @maheakdave you can proceed with Decision Trees. I am assigning you this issue. Do remember to follow the folder structure. Your sub folder titled `Decision Tree Implementation` should be within the folder `MLSelfImplementedAlgos` folder.

**Comment by ArmaanSeth** [comment]
@diffrxction can I work on NLP?

**Comment by diffrxction** [comment]
> @diffrxction can I work on NLP?

You definitely can if you are interested in starting with the very basics. First of all create a new issue in the `Issues` tab with proper title and description of the kind of work you wish to do. Then just follow the same procedure. 

Make a new independent folder called `NLP Basics` in your local fork and add your files within that folder.

**Comment by diffrxction** [comment]
> Hi @diffrxction shouldn't I do it in the issue of basic machine learning algorithms? Or if not, will the new issue that I create have the hactoberfest label.

I think NLP can be treated as its own seperate subset of ML but it definitely is not a basic concept, that's why the separate issue. 

Don't worry about the labels, I will add them once you create the Issue.

**Comment by ArmaanSeth** [comment]
Okay I'll start working on it right away.

**Comment by ArmaanSeth** [comment]
Hi @diffrxction I have created the issue please add the appropriate labels on them.

**Comment by diffrxction** [comment]
> Hi @diffrxction I have created the issue please add the appropriate labels on them.

Done.

**Comment by ArmaanSeth** [comment]
> > Hi @diffrxction I have created the issue please add the appropriate labels on them.
> 
> Done.

Thanks

**Comment by ArmaanSeth** [comment]
Hi @diffrxction, can I implement KNN.

**Comment by diffrxction** [comment]
> Hi @diffrxction, can I implement KNN.

Yeah sure @ArmaanSeth. Put it into a new folder under `MLSelfImplementedAlgos` and add relevant code and documentation as always.

**Comment by diffrxction** [comment]
> > > Hi @diffrxction, can I implement KNN.
> > 
> > 
> > Yeah sure @ArmaanSeth. Put it into a new folder under `MLSelfImplementedAlgos` and add relevant code and documentation as always.
> 
> Okay I'll start working on it and try to finish it till tomorrow.

Sure buddy, no hurry. Take your time.

**Comment by kitrak-rev** [comment]
Hi, I would like to contribute to the Mean Absolute Error.

**Comment by Padi-Rishitha** [comment]
@diffrxction Can I work on mean squared error !!

**Comment by diffrxction** [comment]
> @diffrxction Can I work on mean squared error !!

Sure @Padi-Rishitha. Feel free to go ahead with this task. You can create a folder called `Error Metrics` within the pre-existing `MLSelfImplementedAlgos` folder, the same as I have mentioned in my comment above. And add your code within this folder structure. 

Both code and documentation would be better. And please remember to name the folders in the same way as I have mentioned or they will not be in sync with any future or past commits and show up elsewhere.

I am assigning you this issue. Good luck!

**Comment by kitrak-rev** [comment]
Can I work on Linear Regression?

**Comment by diffrxction** [comment]
> @diffrxction I am done with Mean Absolute Error, kindly look into the PR #29.

Done. Thanks for your contribution!

**Comment by kitrak-rev** [comment]
@diffrxction  Can I do Root Mean square error metric?
